File: Sorting_Algorithm/JAVA/QuickSort.java
package Sorting_Algorithm;
import java.util.*;
public class QuickSort{
    public static int partition(int arr[],int s,int e){
            int pivotelement=arr[s];
            int count=0;
            for(int i=s+1;i<=e;i++){
                if(arr[i]<pivotelement){
                    count++;
                }
            }
            int temp=arr[s+count];
            arr[s+count]=arr[s];
            arr[s]=temp;
            int i=s;
            int j=e;
            while(i<j){
                if(arr[i]<pivotelement){
                    i++;
                }
                else if(arr[j]>=pivotelement){
                    j--;
                }
                else{
                    temp=arr[i];
                    arr[i]=arr[j];
                    arr[j]=temp;
                    i++;
                    j--;
                }
            }
            return s+count;
    }
    public static void quicksort(int[]arr,int s,int e){
        if(s>=e){
            return;
        }
        int pivotindex=partition(arr,s,e);
        quicksort(arr, s, pivotindex-1);
        quicksort(arr, pivotindex+1, e);
    }
    public static void main(String[]args){
        int[]arr={9,8,7,6,5,4,3,2,1};
        quicksort(arr, 0, arr.length-1);
        System.out.println(Arrays.toString(arr));
    }
}

File: Sorting Algorithms (Recursion)/QuickSort.cpp
#include<iostream>
using namespace std;

int partition(int* arr, int s, int e){

    int temp = arr[s];
    int count = 0;
    for (size_t x = s+1; x <= e; x++)
    {
        if(arr[x] <= temp) count++;
    }
    int pivot = s + count;
    // cout<< arr[pivot] << " ";
    swap(arr[pivot],arr[s]);
    int i = s, j = e;
    while(i < pivot && j > pivot){
        while(arr[i] <= arr[pivot]) 
        {
            i++;
        }

        while(arr[j] > arr[pivot]) {
            j--;
        }
        if(i < pivot && j > pivot){
             swap(arr[i++], arr[j--]);
        } 
    } 
    return pivot;
}

void QuickSort(int* arr, int s, int e){

    if(s>=e) return;

    int p = partition(arr,s,e);
    // cout << endl;

    QuickSort(arr,s,p-1);
    QuickSort(arr,p+1,e);
}


int main(int argc, char const *argv[])
{
    int arr[] = {1,5,3,8,6,9,4};
    for (size_t i = 0; i < 7; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl << endl;
    QuickSort(arr, 0, 6);
    for (size_t i = 0; i < 7; i++)
    {   
        cout << arr[i] << " ";
    }
    return 0;
}